### ISSUE TITLE
Updating scene.visuals.Tube's set_data with xyz coordinates fails.

### ISSUE THREAD

**Issue by lucasnavisens** [vispy/vispy#1452]
It seems that use Tube object doesn't have a set_data method and uses directly its parent object
(visual.Mesh) set_data which is expecting vertices instead of xyz coordinates (basically I guess it doesn't handle the xyz->vertices conversion to create a Tube).

Here is some code from the example which works:
l1 = scene.visuals.Tube(points1,
                        shading='flat',
                        color=colors,  # this is overridden by
                        vertex_colors=vertex_colors,
                        tube_points=8)

Just by adding:
l1.set_data(points1)

Which is basically just setting the data to the same array of data,
which previously worked in the constructor.

It fails with the error:
ValueError: incompatible dimensions for cross product (dimension must be 2 or 3)

**Comment by kmuehlbauer** [comment]
@lucasnavisens Correct, the set_data-method is not overridden in Tube-Visual. We can assist you in adding it. PR welcome. Want to give it a try?

**Comment by lucasnavisens** [comment]
Hey!

Great thanks for the opportunity to contribute but have never gotten around to doing it.

I will see in the future if I can contribute !

Cheers.

🥖

**Comment by Pooppap** [comment]
Hi,
I hope you guys won't mind that I hijack the thread. Long story short, I managed to refactor the `set_data` method into the `TubeVisual` class; however, the performance tank is in the tube animation. I hope to get advice from you guys on improving the animation performance before I submit a pull request. Here is the link to my fork: https://github.com/Pooppap/vispy/tree/Tube_set_data.
Cheers!

**Comment by aganders3** [comment]
@Pooppap can you open this as a PR in your own fork to make it easier to review and discuss parts of the code?

**Comment by Pooppap** [comment]
Sure! Here is the link to the PR: https://github.com/Pooppap/vispy/pull/1
Also, I see a similar PR already in work: https://github.com/vispy/vispy/pull/2442 -- maybe we can combine this?

**Comment by aganders3** [comment]
> Also, I see a similar PR already in work: https://github.com/vispy/vispy/pull/2442 -- maybe we can combine this?

After looking at both, your changes are quite similar. It looks like #2442 was almost complete, save for adding some tests. Do you want to start with that branch and add some tests?

**Comment by Pooppap** [comment]
Hmm... Working on this would be an honour, but I think I am not well-versed enough in VisPy and OpenGL, so I'll pass this time. I might revisit this after I get the hang of the Visby library and this issue is still open.
